lib/rust/profiler/src/lib.rs

```diff
@@ -431,21 +531,165 @@ pub enum Event {
         /// When the event occurred.
         timestamp: Timestamp,
     },
+    /// Metadata: wrapper with dependency-injected contents.
+    Metadata(Metadata<M>),
+}
+
+impl<M, L> Event<M, L> {
+    /// Produce a new event that may have a different metadata type, with metadata values
+    /// converted by the given function.
+    fn map_metadata<F, M1>(self, mut f: F) -> Event<M1, L>
+    where F: FnMut(M) -> M1 {
+        match self {
+            Event::Start(start) => Event::Start(start),
+            Event::StartPaused(start) => Event::StartPaused(start),
+            Event::End { id, timestamp } => Event::End { id, timestamp },
+            Event::Pause { id, timestamp } => Event::Pause { id, timestamp },
+            Event::Resume { id, timestamp } => Event::Resume { id, timestamp },
+            Event::Metadata(Metadata { timestamp, data }) =>
+                Event::Metadata(Metadata { timestamp, data: f(data) }),
+        }
+    }
+}
+
+
+
+// ================
+// === Metadata ===
+// ================
+
+/// Wrapper adding a timestamp to dependency-injected contents.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct Metadata<M> {
+    /// When the event occurred.
+    pub timestamp: Timestamp,
+    /// The data.
+    pub data:      M,
 }
 
+impl<M> From<Metadata<M>> for Event<M, &'static str> {
+    fn from(metadata: Metadata<M>) -> Self {
+        Event::Metadata(metadata)
+    }
+}
+
+
+// === OpaqueMetadata ===
+
+/// Black-box metadata object, for use when metadata doesn't need to be created/interpreted.
+#[derive(Debug, Copy, Clone, serde::Serialize, serde::Deserialize)]
+pub enum OpaqueMetadata {
```

Comment:
Could you extend these docs slightly, please? What I'm missing is info what use cases does it cover – when the metadata doesn't need to be interpreted?

Reply:
Used only for testing now, so I made it a private item in `tests`.